### PR TITLE
Types cleanup2

### PR DESCRIPTION
**Pull request by rbuckton**
* Adds a new `ThisType` syntax kind, which will be needed by tree transforms to disambiguate `this` in a type position from `this` in an expression position.
* Adds minimal support for `this` types in conjunction with `--emitDecoratorMetadata`. It now prints "Object" rather than crashing the compiler with a debug assertion.
  * NOTE: This will probably change from "Object" to at least the class constructor for the method after we fix circular reference issues with decorator metadata.

### CODE REVIEW

**Comment by mhegazy** [comment]
:+1:

File: src/compiler/types.ts [hunk 1]
```diff
@@ -726,6 +727,7 @@ namespace ts {
     // @kind(SyntaxKind.StringKeyword)
     // @kind(SyntaxKind.SymbolKeyword)
     // @kind(SyntaxKind.VoidKeyword)
+    // @kind(SyntaxKind.ThisType)
```
Comment (DanielRosenwasser) [inline src/compiler/types.ts]: What about string literal types?

Comment (weswigham) [inline src/compiler/types.ts]: I think there would need to be a `StringLiteralType` kind and a `StringLiteral` kind, just like how there's `ThisType` and `ThisKeyword`. So maybe a followup PR?

File: src/compiler/emitter.ts [hunk 2]
```diff
@@ -7868,6 +7869,8 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
                     case SyntaxKind.GetAccessor:
                     case SyntaxKind.SetAccessor:
                         return emitAccessor(<AccessorDeclaration>node);
+                    // case SyntaxKind.ThisType:
```
Comment (sandersn) [inline src/compiler/emitter.ts]: This looks like it's left over from development.

File: src/compiler/emitter.ts [hunk 3]
```diff
@@ -2651,7 +2651,7 @@ var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, Promi
             }
 
             function emitCallTarget(node: Expression): Expression {
-                if (node.kind === SyntaxKind.Identifier || node.kind === SyntaxKind.ThisKeyword || node.kind === SyntaxKind.SuperKeyword) {
+                if (node.kind === SyntaxKind.Identifier || node.kind === SyntaxKind.ThisKeyword || /*node.kind === SyntaxKind.ThisType ||*/ node.kind === SyntaxKind.SuperKeyword) {
```
Comment (sandersn) [inline src/compiler/emitter.ts]: should this check be for ThisKeyword or ThisType? Or is the commented ThisType left over from development?

**Comment by sandersn** [comment]
:+1:

File: src/compiler/types.ts (hunk 1 above)
Comment (rbuckton) [inline src/compiler/types.ts]: @DanielRosenwasser, @weswigham: I primarily need this to disambiguate `this` during tree transformations. I have no such need for disambiguation for string literal types. I may have a follow-up PR for this at some point to achieve consistency, but its not as high of a priority.

File: src/compiler/emitter.ts (hunk 3 above)
Comment (rbuckton) [inline src/compiler/emitter.ts]: Comment is left over.

File: src/compiler/emitter.ts (hunk 2 above)
Comment (rbuckton) [inline src/compiler/emitter.ts]: Yep.

**Pull request merged**